### PR TITLE
segment_ids in flash attention

### PR DESCRIPTION
**Pull request by khatwanimohit**
* Adding segment_ids in flash_attention 
* add unit test for attention

### CODE REVIEW

File: MaxText/layers.py [hunk 1]
```diff
@@ -328,40 +326,49 @@ class MultiHeadDotProductAttention(nn.Module):
 
 
   def apply_attention(self, query, key, value, enable_flash_attention,
-                      attention_bias, dropout_rng, deterministic):
+                      decoder_segment_ids, attention_bias, dropout_rng, deterministic, decode):
     """ Apply Attention
     """
-    if enable_flash_attention:
+    if enable_flash_attention and not decode:
```
Comment (rwitten) [inline MaxText/layers.py:332]: "If enable_flash_attention and decode" should be asserting, probably here if it is unsupported.

File: MaxText/tests/attention_test.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+from jax.sharding import Mesh
+from layers import MultiHeadDotProductAttention
+from jax.sharding import PartitionSpec as P
+import numpy as np
+
+import pyconfig
+import sys
+
+class AttentionTest(unittest.TestCase):
+  """Test for the RoPE implementation """
```
Comment (rwitten) [inline MaxText/tests/attention_test.py:31]: RoPE?

**Review by rwitten — APPROVED** [review]
Broadly looks good just some little notes.

File: MaxText/layers.py (hunk 1 above)
Comment (khatwanimohit) [reply MaxText/layers.py:332]: Yes, decode is not supported with flash_attention

File: MaxText/tests/attention_test.py (hunk 2 above)
Comment (khatwanimohit) [reply MaxText/tests/attention_test.py]: fixed

**Review by rwitten — APPROVED** [review]

**Pull request merged**